File: src/android/CDVTestFairy.java
package com.testfairy;

import android.util.Log;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Locale;
import java.util.TimeZone;

public class CDVTestFairy extends CordovaPlugin {
	private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'.'SSS'Z'", Locale.US);
	private static final TimeZone UTC_TIMEZONE = TimeZone.getTimeZone("UTC");

	static {
		DATE_FORMAT.setTimeZone(UTC_TIMEZONE);
	}

	public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
		if ("begin".equals(action)) {
			String apiKey = args.optString(0);
			TestFairy.begin(cordova.getActivity(), apiKey);
		} else if ("identify".equals(action)) {
			if (args.length() == 1) {
				String correlationId = args.optString(0, "");
				TestFairy.identify(correlationId);
			} else if (args.length() == 2) {
				String correlationId = args.optString(0, "");
				JSONObject traitObject = args.optJSONObject(1);
				if (traitObject == null) {
					TestFairy.identify(correlationId);
				} else {
					HashMap<String, Object> traits = new HashMap<String, Object>();
					for (Iterator<String> iterator = traitObject.keys(); iterator.hasNext();) {
						String key = iterator.next();
						Object valueObject = traitObject.get(key);
						if (valueObject instanceof Integer || valueObject instanceof Float || valueObject instanceof Double) {
							traits.put(key, valueObject);
						} else if (valueObject instanceof String) {
							String value = (String) valueObject;
							Date date = CDVTestFairy.parse(value);
							traits.put(key, date == null ? value : date);
						}
					}
					TestFairy.identify(correlationId, traits);
				}
			}
		} else if ("checkpoint".equals(action)) {
			String name = args.optString(0);
			TestFairy.addEvent(name);
		} else if ("setCorrelationId".equals(action)) {
			String id = args.optString(0);
			TestFairy.setCorrelationId(id);
		} else if ("pause".equals(action)) {
			TestFairy.pause();
		} else if ("resume".equals(action)) {
			TestFairy.resume();
		} else if ("log".equals(action)) {
			String message = args.optString(0);
			TestFairy.log("TFCordova", message);
		} else if ("setServerEndpoint".equals(action)) {
			String url = args.optString(0);
			TestFairy.setServerEndpoint(url);
		} else if ("sendUserFeedback".equals(action)) {
			String feedback = args.optString(0);
			TestFairy.sendUserFeedback(feedback);
		} else if ("stop".equals(action)) {
			TestFairy.stop();
		} else if ("setScreenName".equals(action)) {
			String name = args.optString(0);
			TestFairy.setScreenName(name);
		} else if ("setUserId".equals(action)) {
			String userId = args.optString(0);
			TestFairy.setUserId(userId);
		} else if ("setAttribute".equals(action)) {
			String key = args.optString(0);
			String value = args.optString(1);
			TestFairy.setAttribute(key, value);
		} else if ("pushFeedbackController".equals(action)) {
			TestFairy.showFeedbackForm();
		} else if ("enableCrashHandler".equals(action)) {
			TestFairy.enableCrashHandler();
		} else if ("disableCrashHandler".equals(action)) {
			TestFairy.disableCrashHandler();
		} else if ("enableMetric".equals(action)) {
			String method = args.optString(0);
			TestFairy.enableMetric(method);
		} else if ("disableMetric".equals(action)) {
			String method = args.optString(0);
			TestFairy.disableMetric(method);
		} else if ("enableVideo".equals(action)) {
			String policy = args.optString(0);
			String quality = args.optString(1);
			float framesPerSecond = (float)args.optDouble(2, 0);
			TestFairy.enableVideo(policy, quality, framesPerSecond);
		} else if ("disableVideo".equals(action)) {
			TestFairy.disableVideo();
		} else if ("enableFeedbackForm".equals(action)) {
			String method = args.optString(0, "shake");
			TestFairy.enableFeedbackForm(method);
		} else if ("disableFeedbackForm".equals(action)) {
			TestFairy.disableFeedbackForm();
		} else if ("disableAutoUpdate".equals(action)) {
			TestFairy.disableAutoUpdate();
		} else if ("setMaxSessionLength".equals(action)) {
			float length = (float)args.optDouble(0, 0);
			TestFairy.setMaxSessionLength(length);
		} else if ("logException".equals(action)) {
			String message = args.optString(0);
			String stacktrace = args.optString(1);
			TestFairy.logThrowable(new Exception(message));
		} else if ("showFeedbackForm".equals(action)) {
			String appToken = args.optString(0);
			boolean takeScreenshot = args.optBoolean(1);
			TestFairy.showFeedbackForm(cordova.getActivity(), appToken, takeScreenshot);
		} else if ("installFeedbackHandler".equals(action)) {
			String appToken = args.optString(0);
			TestFairy.installFeedbackHandler(cordova.getActivity(), appToken);
		} else {
			Log.d("TestFairy", "Action " + action + " is not supported on Android.");
			return false;
		}

		callbackContext.success();
		return true;
	}

	private static Date parse(String value) {
		try {
			return DATE_FORMAT.parse(value);
		} catch (Exception exception) {
			return null;
		}
	}
}

File: src/ios/CDVTestFairy.m
#import "CDVTestFairy.h"
#import <TestFairy/TestFairy.h>
#import <CoreLocation/CoreLocation.h>
#import <Cordova/CDVPluginResult.h>

@implementation CDVTestFairy

- (void)begin:(CDVInvokedUrlCommand*)command
{
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = nil;

	if ([arguments count] > 0) {
		NSString *APIKey = [arguments objectAtIndex:0];
		[TestFairy begin:APIKey];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"API Key is missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)pushFeedbackController:(CDVInvokedUrlCommand*)command
{
	[TestFairy pushFeedbackController];
	CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)updateLocation:(CDVInvokedUrlCommand*)command
{
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = nil;

	if ([arguments count] > 0)
	{
		NSDictionary *locationData = [arguments firstObject];
		NSDictionary *coord = [locationData objectForKey:@"coords"];
		CLLocation *location = [[CLLocation alloc] initWithCoordinate:CLLocationCoordinate2DMake([[coord objectForKey:@"latitude"] doubleValue], [[coord objectForKey:@"longitude"] doubleValue])
															 altitude:[[coord objectForKey:@"altitude"] longValue]
												   horizontalAccuracy:[[coord objectForKey:@"accuracy"] longValue]
													 verticalAccuracy:[[coord objectForKey:@"altitudeAccuracy"] doubleValue]
															timestamp:[NSDate date]];
		[TestFairy updateLocation:[NSArray arrayWithObject:location]];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"location is missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)checkpoint:(CDVInvokedUrlCommand*)command
{
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = nil;

	if ([arguments count] > 0) {

		NSString *name = [arguments objectAtIndex:0];
		[TestFairy checkpoint:name];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"name is missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setCorrelationId:(CDVInvokedUrlCommand*)command
{
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = nil;

	if ([arguments count] > 0) {

		NSString *correlationId = [arguments objectAtIndex:0];
		[TestFairy setCorrelationId:correlationId];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"correlation ID is missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)pause:(CDVInvokedUrlCommand*)command
{
	[TestFairy pause];
	CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)resume:(CDVInvokedUrlCommand*)command
{
	[TestFairy resume];
	CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)identify:(CDVInvokedUrlCommand*)command {
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = nil;

	if ([arguments count] <= 0) {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Correlation ID cannot be empty"];
		[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
		return;
	}

	NSString *correlationId = [arguments objectAtIndex:0];
	NSDictionary *traitValues = [command argumentAtIndex:1 withDefault:@{}];
	if ([traitValues count] == 0) {
		[TestFairy identify:correlationId];
	} else {
//		NSLocale *enUSPOSIXLocale = [NSLocale localeWithLocaleIdentifier:@"en_US_POSIX"];
//		NSDateFormatter *_dateFormatter = [[NSDateFormatter alloc] init];
//		[_dateFormatter setLocale:enUSPOSIXLocale];
//		[_dateFormatter setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss'.'SSS'Z'"];
//
//		NSMutableDictionary *traits = [NSMutableDictionary dictionaryWithCapacity:[traitValues count]];
//		for (NSString* key in traitValues) {
//			id value = [traitValues objectForKey:key];
//			NSDate *date = [_dateFormatter dateFromString:value];
//			[traits setObject:(date == nil ? value : date) forKey:key];
//		}
		[TestFairy identify:correlationId traits:traitValues];
	}

	pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)log:(CDVInvokedUrlCommand*)command {
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = nil;

	if ([arguments count] > 0) {
		NSString *message = [arguments objectAtIndex:0];
		TFLog(@"%@", message);
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"message is missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setServerEndpoint:(CDVInvokedUrlCommand*)command {
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = nil;

	if ([arguments count] > 0) {
		NSString *url = [arguments objectAtIndex:0];
		[TestFairy setServerEndpoint:url];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"url is missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)hideWebViewElements:(CDVInvokedUrlCommand*)command {
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = nil;

	if ([arguments count] > 0) {
		NSString *cssSelector = [arguments objectAtIndex:0];
		[TestFairy hideWebViewElements:cssSelector];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"selector is missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)sendUserFeedback:(CDVInvokedUrlCommand *)command {
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = nil;

	if ([arguments count] > 0) {
		NSString *feedback = [arguments objectAtIndex:0];
		[TestFairy sendUserFeedback:feedback];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"feedback is missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)stop:(CDVInvokedUrlCommand *)command  {
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	[TestFairy stop];

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setScreenName:(CDVInvokedUrlCommand *)command  {
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = nil;

	if ([arguments count] > 0) {
		NSString *name = [arguments objectAtIndex:0];
		[TestFairy setScreenName:name];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"screen name is missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setUserId:(CDVInvokedUrlCommand *)command  {
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = nil;

	if ([arguments count] > 0) {
		NSString *userId = [arguments objectAtIndex:0];
		[TestFairy setUserId:userId];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"userId is missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setAttribute:(CDVInvokedUrlCommand *)command  {
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = nil;

	if ([arguments count] > 1) {
		NSString *key = [arguments objectAtIndex:0];
		NSString *value = [arguments objectAtIndex:1];
		[TestFairy setAttribute:key withValue:value];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"key and value are missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)enableCrashHandler:(CDVInvokedUrlCommand*)command {
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	[TestFairy enableCrashHandler];

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)disableCrashHandler:(CDVInvokedUrlCommand*)command {
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	[TestFairy disableCrashHandler];

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)enableMetric:(CDVInvokedUrlCommand*)command {
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	if ([arguments count] > 0) {
		NSString *metric = [arguments objectAtIndex:0];
		[TestFairy enableMetric:metric];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"metric is missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)disableMetric:(CDVInvokedUrlCommand*)command {
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	if ([arguments count] > 0) {
		NSString *metric = [arguments objectAtIndex:0];
		[TestFairy disableMetric:metric];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"metric is missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)enableVideo:(CDVInvokedUrlCommand*)command {
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	if ([arguments count] > 2) {
		NSString *policy = [arguments objectAtIndex:0];
		NSString *quality = [arguments objectAtIndex:1];
		float framesPerSecond = [[arguments objectAtIndex:2] floatValue];
		[TestFairy enableVideo:policy quality:quality framesPerSecond:framesPerSecond];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"arguments are missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)disableVideo:(CDVInvokedUrlCommand*)command {
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	[TestFairy disableVideo];

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)enableFeedbackForm:(CDVInvokedUrlCommand*)command {
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	if ([arguments count] > 0) {
		NSString *method = [arguments objectAtIndex:0];
		[TestFairy enableFeedbackForm:method];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"method is missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)disableFeedbackForm:(CDVInvokedUrlCommand*)command {
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	[TestFairy disableFeedbackForm];

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)disableAutoUpdate:(CDVInvokedUrlCommand*)command {
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	[TestFairy disableAutoUpdate];

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setMaxSessionLength:(CDVInvokedUrlCommand*)command {
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	if ([arguments count] > 0) {
		float length = [[arguments objectAtIndex:0] floatValue];
		[TestFairy setMaxSessionLength:length];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"length is missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)logException:(CDVInvokedUrlCommand*)command {
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	if ([arguments count] > 1) {
		NSString *message = [arguments objectAtIndex:0];
		NSString *trace = [arguments objectAtIndex:1];
		NSError *error = [NSError errorWithDomain:@"com.testfairy.cordova" code:-1 userInfo:@{NSLocalizedDescriptionKey: message}];
		[TestFairy logError:error stacktrace:[trace componentsSeparatedByString:@"\n"]];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"arguments are missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)showFeedbackForm:(CDVInvokedUrlCommand*)command {
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	if ([arguments count] > 1) {
		NSString *appToken = [arguments objectAtIndex:0];
		BOOL takeScreenshot = [[arguments objectAtIndex:1] boolValue];
		[TestFairy showFeedbackForm:appToken takeScreenshot:takeScreenshot];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"arguments are missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)installFeedbackHandler:(CDVInvokedUrlCommand*)command {
	NSArray* arguments = command.arguments;
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

	if ([arguments count] > 0) {
		NSString *appToken = [arguments objectAtIndex:0];
		[TestFairy installFeedbackHandler:appToken method:@"shake|screenshot"];
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
	} else {
		pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"arguments are missing"];
	}

	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

@end
